akka-typed/src/main/scala/akka/typed/internal/ActorCell.scala

```diff
@@ -0,0 +1,433 @@
+          unsafe.getAndAddInt(this, statusOffset, 1) // the first 1 was just the “active” bit, now add 1msg
+          // if the actor was not yet running, set it in motion; spurious wakeups don’t hurt
+          executionContext.execute(this)
+        }
+      }
+    } catch handleException
+
+  def sendSystem(signal: SystemMessage): Unit = {
+    @tailrec def needToActivate(): Boolean = {
+      val currentList = _systemQueue
+      if (currentList.head == NoMessage) {
+        system.deadLetters.toImpl.sendSystem(signal)
+        false
+      } else {
+        unsafe.compareAndSwapObject(this, systemQueueOffset, currentList.head, (signal :: currentList).head) || {
+          signal.unlink()
+          needToActivate()
+        }
+      }
+    }
+    try {
+      if (needToActivate()) {
+        val old = unsafe.getAndAddInt(this, statusOffset, 1)
+        if (isClosed(old)) {
+          // nothing to do
+          if (Debug) println(s"[$thread] $self NOT enqueueing $signal: terminating")
+        } else if (activations(old) == 0) {
+          // all is good: we signaled the transition to active
+          if (Debug) println(s"[$thread] $self enqueueing $signal: activating")
+          executionContext.execute(this)
+        } else {
+          // take back that token: we didn’t actually enqueue a normal message and the actor was already active
+          if (Debug) println(s"[$thread] $self enqueueing $signal: already active")
+          unsafe.getAndAddInt(this, statusOffset, -1)
+        }
+      } else if (Debug) println(s"[$thread] $self NOT enqueueing $signal: terminated")
+    } catch handleException
+  }
+
+  /**
+   * Main entry point into the actor: the ActorCell is a Runnable that is
+   * enqueued in its Executor whenever it needs to run. The _status field is
+   * used for coordination such that it is never enqueued more than once at
+   * any given time, because that would break the Actor Model.
+   *
+   * The idea here is to process at most as many messages as were in queued
+   * upon entry of this method, interleaving each normal message with the
+   * processing of all system messages that may have accumulated in the
+   * meantime. If at the end of the processing messages remain in the queue
+   * then this cell is rescheduled.
+   *
+   * All coordination occurs via a single Int field that is only updated in
+   * wait-free manner (LOCK XADD via unsafe.getAndAddInt), where conflicts are
+   * resolved by compensating actions. For a description of the bit usage see
+   * the companion object’s source code.
+   */
+  override final def run(): Unit = {
+    if (Debug) println(s"[$thread] $self entering run(): interrupted=${Thread.currentThread.isInterrupted}")
+    val status = _status
+    val msgs = messageCount(status)
+    var processed = 0
+    try {
+      unscheduleReceiveTimeout()
+      if (!isClosed(status)) {
+        while (processAllSystemMessages() && !queue.isEmpty() && processed < msgs) {
```

Comment:
Ok, I see here: https://github.com/akka/akka/pull/21128/files#diff-92d9e38d1b6b284e38230047feea5fdcR207
So the activation counter is increased before the actual enqueueing, so it is not guaranteed that we see it here yet.